### DISCUSSION TITLE
Switches for Siren and LED lights.

### DISCUSSION THREAD

**Discussion by m2wall** [fwestenberg/reolink_dev#293] [Ideas]
Hi,

I have the new RLC-822. It has built in LED lights and also a Siren. 
It would be really useful to control these through Home Assistant. Have alarm set off on all cameras when motion in x camera triggers, and so on.

**Comment by gogreenpower** [comment]
I second this, I have a few E1 Outdoors with LEDs ans Sirens I would like exposed to HA as well.

**Comment by ptyers** [comment]
Still learning how to interact with community to allow me to participate in joint contribution but I have forked the reolink module repo and have started to look at adding SetWhiteLed capability. On my Reolink Duo GetWhiteLed returns valid data so state checking should be possible.

Can you please run the following in your terminal 

curl -X POST 'http://yourcameraIP/cgi-bin/api.cgi?user=yourusername&password=yourpassword' -d '[{"cmd": "SetWhiteLed", "param": { "WhiteLed": {"bright" : 100, "channel": 0, "mode": 1, "state": 1}}}]'

should turn on LEDs

qnd

curl -X POST 'http://yourcameraIP/cgi-bin/api.cgi?user=yourusername&password=yourpassword' -d '[{"cmd": "SetWhiteLed", "param": { "WhiteLed": {"bright" : 100, "channel": 0, "mode": 3, "state": 0}}}]'

should turn off LEDs

and

curl -X POST 'http://yourcameraIP/cgi-bin/api.cgi?user=yourusername&password=yourpassword' -d '[{"cmd": "GetWhiteLed", "action": 0 , "param": { "channel": 0 }}]'

should return status of settings

On one of your E1's and let me know what the jason results are please. Haven't looked at siren yet, although Duo has one.

Then the trick is modifying reolink_dev to support the WhiteLed and Siren capability.

**Reply by m2wall** [reply]
@ptyers I tested this on my RLC-811A. The first curl-command worked to turn on the led, also the third worked to get the data. However, for the second did not work but it does work to use the first command turn it off with mode set to 1 and state set to 0. Thanks alot for this. Hoping to see your code in the min branch in the future.

**Comment by ptyers** [comment]
I need some help here

I have made necessary mofications to switch.py to implement both spotlight and siren functions
BUT !!!!!
with the Reolink Duo (twin lens hence two channel) all common controls 
spotlight/siren/irLights/push notification/recording/record audio
are duplicated but status of switches is not linked.

I obviously need to check and suppress regeneration on channel two controls  (or link switch statuses) if channel one exists but as yet am
not familiar enough with the overall reolink_dev/entity definitions to work out how to do this.
Suggestions gratefully accepted.
In the meanwhile should I do a pull request with what I've got ??

**Reply by gogreenpower** [reply]
Can't help with the code but I'm keen to give it a test

**Comment by ptyers** [comment]
I have added a check to hopefully obviate creation of duplicated common controls,
Current status - pull request pending approval to merge

**Reply by cpainchaud** [reply]
@ptyers pull request actually waiting for your comment :p

**Comment by cpainchaud** [comment]
please try 0.40 and let us know

**Reply by m2wall** [reply]
The spotlight works really great but unfortunately siren does not.

**Reply by ptyers** [reply]
checking out the siren problem. Will advise when resolved.

**Reply by ptyers** [reply]
Update:
Have found problem cause I believe. 
When I did my testing I had the camera in same location as I was. This caused it to detect motion regularly. 
Have now discovered that with current code siren will only initiate when motion is detected. Currently investigating how to fake motion detection as part of siren initiation

Given that Reolink App can enable the Siren from screen there must be a way.

**Comment by cpainchaud** [comment]
@ptyers that comment is for you, I don't have a siren to test with!

**Reply by ptyers** [reply]
Reference detection dual/single channel cameras

I explained myself badly obviously.
for any camera single or dual lens the unique id is comprised of the mac address (with intervening colons removed) and a trailing "-"and a channel id. For a single lens camera this channel id is "1" , a dual lens camera will have the same 13 character (12 for mac address plus one for the "-") leading part but will return both a "1" and a "2" as the suffix to the unique id. With an NVR the same will apply same mac address but multiple channel id's. With an NVR the switches for each channel apply to different cameras and thus should not be suppressed.
On a dual lens camera (DUO) the switches are common controls and should not be duplicated. In fact if not suppressed, duplicate  non-synchronised versions are created. On single lens cameras the if statement is effectively a no-op.

I considered a table of models but deemed it non-desirable because it would require constant updating as new models were introduced. Given the current discussions regarding live streaming, where dependent on model, usage of H265 for streaming appears to be model dependent, even when a GetAbilities returns a valid for H265 this may still be necessary and could be retrofitted to the switch generation.

**Reply by Axionkt** [reply]
Since updating my log is spammed with the following message every minute and enabling debug doesn't give more information:

```
2021-12-02 13:28:08 ERROR (MainThread) [reolink.camera_api] Traceback (most recent call last):
  File "/home/homeassistant/homeassistant/lib/python3.9/site-packages/reolink/camera_api.py", line 754, in map_json_response
    self._audio_alarm_state = data["value"]["Audio"]["enable"]
KeyError: 'enable'
```

My camera is a E1 Zoom which does have a siren so I guess this error message is related to this addition.

**Reply by ptyers** [reply]
My apologies. Error on my part. Misread the doco re version of Audio Alarm. Data returned in marginally different json form between ver=0 and Ver=1.
Will endeavour to do a pull request on underlying reolink module tonight.
Ver=0 should read data["value"]["Audio"] ["schedule"]["enable"] 
whereas Ver=1 reads as data["value"]["Audio"]["enable"]
I did all my testing with a Ver=1 camera (DUO) obviously the E1 Zoom must be Ver=0 camera - again my apologies

**Reply by ptyers** [reply]
Pull request correcting, I believe, this bug has been submitted on underlying module reolink.

**Reply by Axionkt** [reply]
No problem, thanks for the pull request, being able to control the siren is a nice addition.
Everything works fine now.

**Comment by cpainchaud** [comment]
Yes,  please try latest version

**Reply by gogreenpower** [reply]
Tested on my E1 Outdoor.

Turn Siren on, nothing.
Turn Siren on, walk in front of camera, siren goes off.
Turn Siren off, siren stops.
Turn Siren off, walk in front of camera, no siren.

**Reply by gogreenpower** [reply]
Version 0.42

**Reply by gogreenpower** [reply]
Errors in Log:

`Logger: homeassistant.components.stream.worker
Source: components/stream/worker.py:487
Integration: Stream (documentation, issues)
First occurred: 5:13:32 PM (13 occurrences)
Last logged: 5:41:47 PM

Error demuxing stream:
Error demuxing stream: [Errno 1414092869] Immediate exit requested: 'rtmp://192.168.1.27:1935/bcs/channel0_main.bcs?channel=0&stream=0&user=*****&password=******'

This error originated from a custom integration.

Logger: custom_components.reolink_dev
Source: helpers/update_coordinator.py:193
Integration: Reolink IP camera (documentation, issues)
First occurred: 5:07:45 PM (5 occurrences)
Last logged: 5:42:02 PM

Timeout fetching reolink data


This error originated from a custom integration.

Logger: homeassistant
Source: custom_components/reolink_dev/base.py:258
Integration: Reolink IP camera (documentation, issues)
First occurred: 5:07:16 PM (151 occurrences)
Last logged: 5:42:52 PM

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/reolink_dev/sensor.py", line 98, in _update_event_range
    search, _ = await self._base.send_search(start, end, True)
  File "/config/custom_components/reolink_dev/base.py", line 258, in send_search
    return await self._api.send_search(start, end, only_status)
  File "/usr/local/lib/python3.9/site-packages/reolink/camera_api.py", line 1445, in send_search
    response = await self.send(body, {"cmd": command})
  File "/usr/local/lib/python3.9/site-packages/reolink/camera_api.py", line 1555, in send
    async with self._aiohttp_session.post(
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client.py", line 1117, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client.py", line 544, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 905, in start
    self._continue = None
  File "/usr/local/lib/python3.9/site-packages/aiohttp/helpers.py", line 656, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError


This error originated from a custom integration.

Logger: custom_components.reolink_dev.binary_sensor
Source: custom_components/reolink_dev/binary_sensor.py:115
Integration: Reolink IP camera (documentation, issues)
First occurred: 5:07:48 PM (159 occurrences)
Last logged: 5:42:52 PM

Motion states could not be queried from API


This error originated from a custom integration.

Logger: homeassistant
Source: custom_components/reolink_dev/binary_sensor.py:116
Integration: Reolink IP camera (documentation, issues)
First occurred: 5:07:48 PM (159 occurrences)
Last logged: 5:42:52 PM

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/reolink_dev/binary_sensor.py", line 112, in handle_event
    await self._base.api.get_all_motion_states()
  File "/usr/local/lib/python3.9/site-packages/reolink/camera_api.py", line 585, in get_all_motion_states
    response = await self.send(body)
  File "/usr/local/lib/python3.9/site-packages/reolink/camera_api.py", line 1555, in send
    async with self._aiohttp_session.post(
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client.py", line 1117, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client.py", line 544, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 905, in start
    self._continue = None
  File "/usr/local/lib/python3.9/site-packages/aiohttp/helpers.py", line 656, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/reolink_dev/binary_sensor.py", line 116, in handle_event
    _LOGGER.error(traceback.format_exc(()))
  File "/usr/local/lib/python3.9/traceback.py", line 167, in format_exc
    return "".join(format_exception(*sys.exc_info(), limit=limit, chain=chain))
  File "/usr/local/lib/python3.9/traceback.py", line 120, in format_exception
    return list(TracebackException(
  File "/usr/local/lib/python3.9/traceback.py", line 495, in __init__
    context = TracebackException(
  File "/usr/local/lib/python3.9/traceback.py", line 517, in __init__
    self.stack = StackSummary.extract(
  File "/usr/local/lib/python3.9/traceback.py", line 340, in extract
    if limit >= 0:
TypeError: '>=' not supported between instances of 'tuple' and 'int'
`

**Reply by ptyers** [reply]
Still investigating how to cause siren to be àctivated by switch without camera actually detecting motion.

The log entries you included appear to relate to the search function  the changes I made for siren/spotlight should have had no effect on the search function, entirely separate part of module.

**Comment by ptyers** [comment]
Have found reference to an API AudioAlarmPlay but the reference had parameter problem. Am investigating, this should get around the siren only activating when motion detected.
Update:

Can owners of any Reolink camera with siren capability please try the following and get back to me here with results (replacing capitalised items IP, USER,PASSWORD with local parameters)

TURN ON SIREN
$ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "manul", "manual_switch": 1, "times": 2, "channel": 0}}]'

TURN OFF SIREN
$ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "manul", "manual_switch": 0, "times": 2, "channel": 0}}]'

TURN ON SIREN FOR 3 CYCLES THEN STOP
$ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "times", "manual_switch": 1, "times": 3, "channel": 0}}]'

Depending on reported results I will create a pull request for the Reolink module to replace existing (not so good attempt) at siren initiation. This will be relflected in the Reolink_dev addon next update.

**Reply by Axionkt** [reply]
Here is the response I get with the E1 Zoom
[
   {
      "cmd" : "Unknown",
      "code" : 1,
      "error" : {
         "detail" : "not support",
         "rspCode" : -9
      }
   }
]

**Reply by cpainchaud** [reply]
@ptyers could there be a V20 API on this one? :)

**Reply by ptyers** [reply]
cpainchaud: There does not appear (stress so far, does not appear) to be a V20 version). I tried AudioAlarmPlayV20) and got back an unrecconised

Axionkt - using the Reolink App can you activate the siren as part of Audio settings? 
From the web description the E1 Zoom has Audio but it is not clear if there is a siren capability.

Also can you run the following and attach the result file here. Results will be sent to text file not the screen.
Note that first curl requires multiple replacement of USERNAME

$ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd": "GetAbility","action": 1, "param": {"User": {"username": USERNAME}}}]' > E1Results.txt

$ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd": "GetAudioAlarm", "action": 1, "param": {"channel": 0}},]' >> E1Results.txt

$ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd": "GetAudioAlarmV20", "action": 1, "param": {"channel": 0}}]' >> E1Results.txt

**Reply by m2wall** [reply]
> Have found reference to an API AudioAlarmPlay but the reference had parameter problem. Am investigating, this should get around the siren only activating when motion detected. Update:
> 
> Can owners of any Reolink camera with siren capability please try the following and get back to me here with results (replacing capitalised items IP, USER,PASSWORD with local parameters)
> 
> TURN ON SIREN $ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "manul", "manual_switch": 1, "times": 2, "channel": 0}}]'
> 
> TURN OFF SIREN $ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "manul", "manual_switch": 0, "times": 2, "channel": 0}}]'
> 
> TURN ON SIREN FOR 3 CYCLES THEN STOP $ curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "times", "manual_switch": 1, "times": 3, "channel": 0}}]'
> 
> Depending on reported results I will create a pull request for the Reolink module to replace existing (not so good attempt) at siren initiation. This will be relflected in the Reolink_dev addon next update.

The two first commands work with  RLC-811A (POE-camera) 
however the 3rd, the cycle one, seems to just keep go on.

**Reply by ptyers** [reply]
I will recheck against my DUO re cycle not ceasing after number requested. Only did it for times:2  rather than 3. Each cycle is about 8 seconds

**Reply by Axionkt** [reply]
@ptyers here are the results, and yes I can trigger the siren in the app.
[E1ZoomResults.txt](https://github.com/fwestenberg/reolink_dev/files/7668344/E1ZoomResults.txt)

**Reply by m2wall** [reply]
> I will recheck against my DUO re cycle not ceasing after number requested. Only did it for times:2 rather than 3. Each cycle is about 8 seconds

Let me know if you want me to try again. My mom came out looking confused when it happened. Im setting this up for her while we're living in a small house next to hers waiting for our house to get built :p

**Reply by ptyers** [reply]
axionkt
My apologies I seem to have mistyped the first script lets try again - Note that USERNAME inside curly brackets is quoted
I expect one of the latter two to fail but not both

curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd": "GetAbility","action": 1, "param": {"User": {"userName": "USERNAME"}}}]' > E1Results.txt

curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd": "GetAudioAlarm", "action": 1, "param": {"channel": 0}}]' >> E1Results.txt

curl -X POST 'http://IP_ADDRESS/cgi-bin/api.cgi?user=USERNAME&password=PASSWORD' -d '[{"cmd": "GetAudioAlarmV20", "action": 1, "param": {"channel": 0}}]' >> E1Results.txt

**Reply by ptyers** [reply]
Tungval - Don't retry just yet.
I'll do further research here regarding cycles and also audio level
UPDATE:
I have tried using the cycle script for various number of cycles 1, 2, 3, 5 and on my DUO it does as expected and turns off after the requested number of cycles. Hang off for the moment until further data from other cameras is in.
As well, its unlikely that there will be a use case for setting siren to run for X cycles from HA. If someone can see a use case let me know, its not hard to implement with the newer code awaiting on the pull request.

**Reply by Axionkt** [reply]
@ptyers My Bad, I missed that when reviewing the requests.
Here are the results, it seems there is no V20 API.
[E1ZoomResults.txt](https://github.com/fwestenberg/reolink_dev/files/7682363/E1ZoomResults.txt)

**Comment by gogreenpower** [comment]
The E1 Outdoor siren can be triggered from the app


________________________________

**Reply by ptyers** [reply]
That's a problem. There may indeed be a V20 version  of AudioAlarmPlay if on the E1 it returns an error state using the POST  data I asked to be tested but from the Reolink app can be Initiated.

Gogreenpower can you execute the scripts against your E1 as confirmation of the error return please.

**Reply by gogreenpower** [reply]
Here are my results,

[e1results1.txt](https://github.com/fwestenberg/reolink_dev/files/7672211/e1results1.txt)

[e1results2.txt](https://github.com/fwestenberg/reolink_dev/files/7672212/e1results2.txt)

[e1results3.txt](https://github.com/fwestenberg/reolink_dev/files/7672214/e1results3.txt)

**Reply by ptyers** [reply]
Thanks Greenpower.'At least I know the scripts to get abilities and that E1 recognises GetAudioAlarm and GetAudioAlarmV20 .

However
Sorry, I didnt make myself clear I meant the three scripts re Turn ON siren, TURN OFF siren, CYCLE siren

I will analyse the results you have provided and endeavour to see if there is a foible with E1 camera. But I'd like confirmation that the AudioAlarmPLay command returns an error with the E1.

**Reply by gogreenpower** [reply]
```
curl -X POST 'http://192.168.1.31/cgi-bin/api.cgi?user=******&password=*********' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "manul", "manual_switch": 1, "times": 2, "channel": 0}}]'
[
   {
      "cmd" : "AudioAlarmPlay",
      "code" : 0,
      "value" : {
         "rspCode" : 200
      }
   }
]
```

Works well, was on for over 30s before I used the second command to turn off.


```
curl -X POST 'http://192.168.1.31/cgi-bin/api.cgi?user=******&password=*******' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "manul", "manual_switch": 0, "times": 2, "channel": 0}}]'
[
   {
      "cmd" : "AudioAlarmPlay",
      "code" : 0,
      "value" : {
         "rspCode" : 200
      }
   }
]

```
The three times one only stays on for approxiamtely 6 seconds.

```
curl -X POST 'http://192.168.1.31/cgi-bin/api.cgi?user=*******&password=**********' -d '[{"cmd":"AudioAlarmPlay","action": 0, "param": {"alarm_mode": "times", "manual_switch": 1, "times": 3, "channel": 0}}]'
[
   {
      "cmd" : "AudioAlarmPlay",
      "code" : 0,
      "value" : {
         "rspCode" : 200
      }
   }
]

```

**Reply by ptyers** [reply]
I think this indicates that E1 will be usable with current code. Not certain why the initial results from axionkt indicated otherwise.
The only potential problem now is whether there is a, as yet, unidentified API which can affect Siren Volume.

**Reply by gogreenpower** [reply]
Thanks for your efforts. I'm happy with siren and led on/off. Anything else is just a bonus for me.

**Comment by m2wall** [comment]
I tried the spotlights a bit more. When i turn it on it appras it triggers the same or something similar to "motion detected". 
The event that the camera uses to trigger if a person is detected. What i mean by this is it turns on the camera but it turns off after a while.

I tried changing the setting on lights from automated (when triggered) to off in hopes it wouldnt interfere with some automated "no motion so turn off spotlight" but it seems after i trigger the spotlight through home assistant it sets the setting back to automated in the camera settings awell. 

I did the same with "always on between x and x" but then its set to off in home assistant even though lights are on and it turns the lights off and togges HA switch on. So lights are off but HA switch is on. This also sets it back to Automatic.

**Reply by ptyers** [reply]
I will have a closer look at this.
At present the turn Spotlight ON just sets the camera to Night Mode ON, AUTO ON, 100% bright which will override whatever was previously set on the camera.

Initial tests indicate an interim fix I think. NIGHT MODE ON, SCHEDULE ON, ENABLE ON, BRIGHT 100% seems to stop the auto off which happens when motion is detected then motion stops. This will override  the camera settings  in a state where spotlight will come on at scheduled times, will not automatically be triggered by motion. Would setting the camera back to NIGHT MODE ON, AUTO ON be acceptable when manually turning Spotlight OFF? Or should I capture state of WhiteLed prior to turning it ON and after turning it off manually restore those old settings. My API doco does not have info on restoring the schedule times but other states can be restored.

UPDATE:
I think I may have it beaten. I have got access to a later version of API documentation which shows how to set LightingSchedule. If I set the WhiteLed to NIGHTMODE ON, ENABLE On, AUTO OFF and schedule to 0000 to 2359 I have now had Spotlight running for over one hour without it turning off. Apparently no longer sensitive to motion or lack thereof.

**Reply by m2wall** [reply]
Graet to hear @ptyers and nice work! You are the best!

**Comment by ptyers** [comment]
Changes have been implemented and merged.
I suggest this discussion be closed and problems reported as issues.
Spotlight will leave camera in Night_mode Auto when turned off
Siren will be left as auto, off when turned off
